tests/test_schema.py

```diff
@@ -0,0 +1,351 @@
+        self.schema_obj.load_schema()
+
+    def test_save_schema(self):
+        """ Try to save a schema """
+        # Load the template schema
+        self.schema_obj.schema_filename = self.template_schema
+        self.schema_obj.load_schema()
+
+        # Make a temporary file to write schema to
+        tmp_file = tempfile.NamedTemporaryFile()
+        self.schema_obj.schema_filename = tmp_file.name
+        self.schema_obj.save_schema()
+
+    def test_load_input_params_json(self):
+        """ Try to load a JSON file with params for a pipeline run """
+        # Make a temporary file to write schema to
+        tmp_file = tempfile.NamedTemporaryFile()
+        with open(tmp_file.name, 'w') as fh:
+            json.dump({'reads': 'fubar'}, fh)
+        self.schema_obj.load_input_params(tmp_file.name)
+
+    def test_load_input_params_yaml(self):
+        """ Try to load a YAML file with params for a pipeline run """
+        # Make a temporary file to write schema to
+        tmp_file = tempfile.NamedTemporaryFile()
+        with open(tmp_file.name, 'w') as fh:
+            yaml.dump({'reads': 'fubar'}, fh)
+        self.schema_obj.load_input_params(tmp_file.name)
+
+    @pytest.mark.xfail(raises=AssertionError)
+    def test_load_input_params_invalid(self):
+        """ Check failure when a non-existent file params file is loaded """
+        self.schema_obj.load_input_params('fubar')
+
+    def test_validate_params_pass(self):
+        """ Try validating a set of parameters against a schema """
+        # Load the template schema
+        self.schema_obj.schema_filename = self.template_schema
+        self.schema_obj.load_schema()
+        self.schema_obj.input_params = {'reads': 'fubar'}
+        assert self.schema_obj.validate_params()
+
+    def test_validate_params_fail(self):
+        """ Check that False is returned if params don't validate against a schema """
+        # Load the template schema
+        self.schema_obj.schema_filename = self.template_schema
+        self.schema_obj.load_schema()
+        self.schema_obj.input_params = {'fubar': 'reads'}
+        assert not self.schema_obj.validate_params()
+
+    def test_validate_schema_pass(self):
+        """ Check that the schema validation passes """
+        # Load the template schema
+        self.schema_obj.schema_filename = self.template_schema
+        self.schema_obj.load_schema()
+        self.schema_obj.validate_schema()
+
+    @pytest.mark.xfail(raises=AssertionError)
+    def test_validate_schema_fail_notjsonschema(self):
+        """ Check that the schema validation fails when not JSONSchema """
+        self.schema_obj.schema = {'type': 'invalidthing'}
+        self.schema_obj.validate_schema()
+
+    @pytest.mark.xfail(raises=AssertionError)
+    def test_validate_schema_fail_nfcore(self):
+        """ Check that the schema validation fails nf-core addons """
+        self.schema_obj.schema = {}
+        self.schema_obj.validate_schema()
```

Comment:
isn't that only a test for a failed validation of an empty schema?

Reply:
Yup - although I think an empty JSON object is probably valid JSON Schema, we test to make sure that it has at least a `properties` key. So this should raise that nf-core specific `AssertionError`